Use UI Queue runner for import

eileenmcnaughton opened the pull request:
Overview
----------------------------------------
Use UI Queue runner for import

Before
----------------------------------------
Faux ajax still times out the proxy setting

After
----------------------------------------
![image](https://user-images.githubusercontent.com/336308/171649754-462da7c4-9179-44c6-bb47-8430c31a0098.png)


Technical Details
----------------------------------------
I got it working in the UI - can test later via CLI

Comments
----------------------------------------
This has a really long chain - most are less scary than they look. The ones that set the path for the rest are the ones that are separately open as PRs with tests

eileenmcnaughton committed:
Use UI Queue runner for import

Additional minor code simplification

Tear down fix

Maybe the var name needs to match in the CI php version

totten commented on CRM/Import/Parser.php:
```diff
@@ -646,6 +646,75 @@ protected function doPostImportActions() {
+      'errorMode' => CRM_Queue_Runner::ERROR_ABORT,
+      'onEndUrl' => CRM_Utils_System::url('civicrm/import/contact/summary', ['user_job_id' => $this->getUserJobID(), 'reset' => 1]),
+    ]);
+    $runner->runAllViaWeb();
```
So in a unit-test context, `$runner->runAll()` would be more appropriate than `$runner->runAllViaWeb()`. Maybe find a conditional/variable to pick the better runner??

eileenmcnaughton replied:
@totten ok - I just looked & yes - that is the cause of the test fails - it didn't work for me in the web without that though - the specific tests ACTUALLY test the form flow - so I guess 

1) the running of it should be moved back to the form
2) the task to set up the queue should take some parameters to make it easier to offer backend as well as front end
3) the tests need to catch the premature exit exception & run the rest of the queue themselves
4) side note - in the csv api explorer (and in fact in the wmf custom import implementation - which is on it's way out) - batch size is a configuration option

![image](https://user-images.githubusercontent.com/336308/171749363-143b0639-83c6-48de-9076-38ffb8987301.png)

eileenmcnaughton replied:
Ok - so what I'm thinking to try in conjunction with your PR is 

- if a queue job already exists with a certain name (civicrm_executor or something genecidal like that) THEN offer the option to process in the background
- in this scenario we would create a queue - the same as now with a background config (error handling  'abort' ) and we would ALSO add an item to the pre-existing `civicrm_executor` to run that queue -(error handling  'discard' )
- the tricky part is I'd need 'some sort of UI' that the user could be redirected to to see what is happening because they are still 'responsible' for checking it completes at this stage.

That's the flow I figure I'll try out @totten

totten replied:
Oh, the batch size option sounds like a good idea.

> the task to set up the queue should take some parameters to make it easier to offer backend as well as front end

Yeah, I think its good practice try to decouple the "build queue" and "run queue" parts. A few other examples:

* If you look at the core-upgrader, it knows how build a queue of upgrade tasks (`CRM_Upgrade_Form::buildQueue()`). Then there are a few frontends:
    * The web page `civicrm/upgrade` calls `buildQueue()` + `runAllViaWeb()`
    * The CLI `drush civicrm-upgrade-db` calls `buildQueue()` + `runAll()`
    * The CLI `cv upgrade:db` calls `buildQueue()` and its own `ConsoleQueueRunner` (*which just shows more colorful output and which respects the `-v`erbose flag*).
* If you look at ext-upgrader, it knows how to build a queue of upgrade tasks (`CRM_Extension_Upgrades::createQueue()`). Then there are a few frontends:
    * The web page `civicrm/admin/extensions/upgrade` calls `createQueue()` + `runAllViaWeb()`
    * The API `Extension.upgrade` calls `createQueue()` + `runAll()`.

totten replied:
> * if a queue job already exists with a certain name (civicrm_executor or something genecidal like that) THEN offer the option to process in the background
> * in this scenario we would create a queue - the same as now with a background config (error handling 'abort' ) and we would ALSO add an item to the pre-existing civicrm_executor to run that queue -(error handling 'discard' )

OK, so for the current PR, the original scope is spot-on (ie *get it working with the foreground AJAX runner*). This is going into follow-up work, right?

FWIW, `coworker` doesn't need a `civicrm_executor` like that -- its standard behavior is to pick up on any queue which matches `runner IS NOT NULL and status = 'active'`.

I would note that `civicrm_queue` has some options that you might call *obscure tunables* (eg `lease_time`, `retry_limit`, `retry_interval`) -- things which should have safe defaults, but which sysadmins might adjust in some edge-cases. A *queue template* could serve as both configuration-element and flag-record. Ex:

```php
// Register queue-template
\Civi\Api4\Queue::create()
  ->setValues([
    'name' => 'import/template',
    'type' => 'SqlParallel',
    'runner' => 'task',
    'status' => 'template', /* ?? OR: 'is_template' => TRUE, ?? */
    'error' => 'abort',
    'lease_time' => 5*60,
    'retry_limit' => 0,
  ]),
  ->execute();
```

Then when it's time to prepare queued tasks, conditionally copy the template:

```php
if (queue('import/template') is an active template) {
  // Found template. Run in background.
  $q = Civi::queue("import/{$jobId}", ['tpl' => 'import/template', 'status' => 'draft']);
  Importer::buildQueue($q);
  $q->setStatus('active'); /* Background worker will take over */
}
else {
  // No template. Run in foreground.
  $q = Civi::queue("import/{$jobId}", ['type' => 'Sql', 'error' => 'abort']);
  Importer::buildQueue($q);
  $r = new CRM_Queue_Runner(...);
  $r->runAllViaWeb();
}
```

eileenmcnaughton replied:
@totten I'm trying to use this use case to also test & QA your PR - so yeah -  I have pulled down your PR & am trying to get it to add the background queue option in conjunction with that - but I've gonna have to have a snooze & come back to it - about 50% of my brain power won't be released until https://github.com/civicrm/civicrm-core/pull/23666 is merged & I can rebase this over it & another  45% seems to be taken up by lack of sleep so if I can chase the rabbit into the cage I'm gonna see if I can take a kip & free up some of the 45%

eileenmcnaughton replied:
& it has been merged - I rebased this

eileenmcnaughton commented:
Same test just passed for me locally - re-running to see impact of merging the Queue PR in the meantime - I might have written to that spec?

I also rebased in a commit that moves around the tpl file a bit  - it is technically simple but part of the contact flow r-run

![image](https://user-images.githubusercontent.com/336308/171779839-3562ca62-1a87-46e7-953d-269a70646c2c.png)

eileenmcnaughton commented:
Those tests are NOT failing for me locally in isolation - I have added to the `tearDown` in case it is that. I also have run the whole class `CRM_Contact_Import_Parser_ContactTest` locally - it's definitely something about test leakage - hopefully the thing I just tried

eileenmcnaughton commented:
Dang - still passing locally - failing in the suite

![image](https://user-images.githubusercontent.com/336308/171821351-ceba2068-3e95-40c1-ab37-5266eab7a7a2.png)

eileenmcnaughton commented:
@totten  this part is passing - I'm very keen to work on the queue issue more tomorrow & keep the momentum over the next few days - but I think we can merge this

eileenmcnaughton commented:
I originally thought this UI bit would be a fairly straight forward merge & once merged we could focus on the background processing - but I think we are treating that as a blocker to merging this - so I've moved some of the parts of this to other PRs so they don't block me cleaning up the other imports - I have rebased that into this PR - hence the rebase & the PR having become 'bigger' again

totten commented:
> (@eileenmcnaughton) I originally thought this UI bit would be a fairly straight forward merge & once merged we could focus on the background processing - but I think we are treating that as a blocker to merging this 

I agree with your original thought -- the option for background processing should be a follow-up issue.

Since this has been rebased a couple times since I last did an `r-run` -- I suppose I should do another (light) `r-run`?  Then merge-on-pass -- and finally (mostly) rely on the RC testers?

eileenmcnaughton commented:
@totten yeah - I think this is mostly about getting it out of the way so we can focus on the background processing issue - I feel like we are close enough to that that we could get it in before we branch & that if we do it will really make a more 'compelling' case for rc testing

As I mentioned @MegaphoneJon has indicated some rc resource from one of his clients

eileenmcnaughton commented:
@totten I've also planning to document how the import flow is working - although I wanted to get the last few entities switched over to it first & then I can - there is a PR up for the membership import & while there are 3 more they become increasingly simply once that is merged

eileenmcnaughton committed:
Remove remaining refs to onDuplicate

eileenmcnaughton commented:
@totten -  the other imports share a lot of code with each other - but not with the contact import

I've pushed up a PR to remove the remaining references to `onDuplicate` as they affect the contact import

eileenmcnaughton committed:
Also remove contactFields from createContact signature

totten commented on CRM/Import/Parser.php:
```diff
@@ -646,6 +656,28 @@ protected function doPostImportActions() {
+    $totalRowCount = $totalRows = $dataSource->getRowCount(['new']);
+    $queue = Civi::queue('user_job_' . $this->getUserJobID(), ['type' => 'Sql', 'error' => 'abort']);
+    $offset = 0;
+    $batchSize = 5;
```
@eileenmcnaughton I assume the small batch size (`5`) is to facilitate inspection (*make it easy to watch the batching mechanism*)? There's a separate/follow-up issue to extract this to a setting? (That's fine - not a blocker.)

But if we wind up with a constant-value in the stable-release -- it should probably be a bit higher (like... 20 or 100 or 500 or 1000...)

eileenmcnaughton replied:
@totten yeah - I think the batch size will be an option on the first page of the import

eileenmcnaughton replied:
I would probably go for around 50 as a default - that should import in 30 seconds on most sites

andyburnsco replied:
Seems like there may be sweet spot for it. 5 I get for testing but doing imports now and it makes it quite slow. I can try witching that hardcoded limit around and see.

eileenmcnaughton replied:
Thanks - let us know what you find

totten replied:
I played around with [some guesstimates about what determines the max batch limit](https://gist.github.com/totten/8f18e04c2198f098521946d6ca3a3222). For example, one might assume that:

* Each PHP worker is limited to 64mb RAM (the baseline required by Civi installer)
* Each HTTP request is bound by the default timeout specified by `apache.org` and `nginx.org` (ie 60s)
    * Per `php.net`, the default timeout for a PHP request is actually lower (30s), but most importers call `set_time_limit(0)`.
* Each import-task requires an average 100ms runtime and 100kb memory.

The gist is a bit fiddly, and (*while fiddling*) the projected limits ranged somewhere around 300 - 600.

For a hard-coded/application-level default, I'd probably vote for 100...

totten commented:
@eileenmcnaughton OK, I've done a bit more `r-run`. It was by no means exhaustive, but I did check some bits like "Add  to existing group" and "Add to new tag" -- and I checked a few other importers (eg `Contributions`, `Activities`). It seems work (at least well as before). There are follow-up tasks/testing, and they don't need to block this PR.

Merge ready.

eileenmcnaughton commented:
Erm - @totten what does merge-ready mean - is that merge-on-pass or we put things on hold for a few days now?

totten commented:
@eileenmcnaughton If you're happy with the grep-results (eg w/onDuplicate update) and if the plan is to address batch-size separately, then it's merge-on-pass.

If there are any more tweaks you want first (like bumping default batch-size to 50), then do that (and then consider it merge-on-pass).

eileenmcnaughton commented:
@totten hmm - OK - re batch size - I could go either way - but my suspicion is it might be easier to test the background processing with it at 5 than 50 & then bump it (if I haven't added a checkbox by then)

eileenmcnaughton committed:
Fix tests to the updated function signature

eileenmcnaughton commented:
Tests all failed because I didn't update the tests to reflect the code was no longer in use - Ive done that now
